Pointer overflow checks should detect overflow in offset multiplication

tautschnig opened the pull request:
Pointer arithmetic requires multiplication of the offset by the size of
the base type (for any base type larger than 1 byte). Such a
multiplication isn't introduced until the back-end, where no opportunity
for adding properties exists anymore. Therefore, synthesize the
multiplication to generate arithmetic overflow checks at the GOTO level.

Fixes: #6631

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Move flag_resett class to the top to enable use in other methods

No changes in behaviour, just code motion. Upcoming commits will use
this class in `check_rec`.

tautschnig committed:
flag_resett: use source_locationt, no need for full instruction

There isn't any use of fields other than the instruction's source
location, so take just that. Also use `as_string()` instead of
`pretty()` for more nicely formatted error reporting.

tautschnig committed:
Pointer overflow checks should detect overflow in offset multiplication

Pointer arithmetic requires multiplication of the offset by the size of
the base type (for any base type larger than 1 byte). Such a
multiplication isn't introduced until the back-end, where no opportunity
for adding properties exists anymore. Therefore, synthesize the
multiplication to generate arithmetic overflow checks at the GOTO level.

Fixes: #6631

chris-ryder commented on src/analyses/goto_check_c.cpp:
```diff
@@ -340,6 +340,82 @@ class goto_check_ct
+///
+/// A flag's initial value (before any `set_flag` or `disable_flag`) is restored
+/// when the entire object goes out of scope.
+class flag_resett
```
I realise this is just code you've moved, but just wondering about the naming of this class (sorry...) Conceptually what this class is really about is less about "resetting" a flag, and more about introducing a "scope" for the flag? So maybe call it something like `flag_scopet` ? or `flag_overridet` ? One for a later PR though.

tautschnig replied:
Good call, I think `flag_overridet` is the best match for what it does. I've inserted a commit to do the renaming.

tautschnig committed:
Rename flag_resett to flag_overridet to be match its behaviour

This class no longer is confined to resetting flags, but now has
additional override capabilities.

Co-authored-by: Chris Ryder <chris.ryder@diffblue.com>

chris-ryder reviewed (approved):
Re-asserting my approval after the extra rename commit, thanks :-)
